File: packages/typescript/src/encodings/formatAggregation.ts
import { a1 } from "../address.ts";
import type {
  Encoding,
  FormatAggregationJson,
  FormatType,
  Grid,
  Origin,
  TokenCounter,
} from "../types.ts";

/**
 * Canonical emission order for format-aggregation groups. The classifier may
 * encounter types in any order; the encoder always emits groups in THIS order
 * (omitting types with no ranges), independent of which row/col surfaced them.
 *
 * Keeping the order fixed here is what makes the encoding deterministic across
 * runs and languages — every port MUST emit groups in the same order.
 */
const TYPE_ORDER: readonly FormatType[] = [
  "IntNum",
  "FloatNum",
  "ScientificNum",
  "PercentageNum",
  "CurrencyData",
  "DateData",
  "TimeData",
  "YearData",
  "EmailData",
  "Boolean",
  "Text",
] as const;

/**
 * Classification patterns, applied in priority order. The first match wins, so
 * more-specific patterns are listed first. Examples that drive the ordering:
 *   - "1900" matches both Year and Int — Year is more specific, must run first
 *   - "1.5e10" matches both Scientific and (loosely) Float — Scientific first
 *   - "$5" matches Currency, not Int — Currency comes before the numeric fallbacks
 */
const BOOLEAN = /^(?:true|false)$/i;
const EMAIL = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
const SCIENTIFIC = /^-?\d+(?:\.\d+)?[eE][+-]?\d+$/;
const PERCENT = /^-?\d+(?:\.\d+)?%$/;
const CURRENCY = /^-?[$€£¥]\d+(?:\.\d+)?$/;
const DATE_ISO = /^\d{4}-\d{1,2}-\d{1,2}$/;
const DATE_SLASH = /^\d{1,2}\/\d{1,2}\/\d{2,4}$/;
const DATE_DASH = /^\d{1,2}-\d{1,2}-\d{2,4}$/;
const TIME_12 = /^\d{1,2}:\d{2}(?::\d{2})?\s?(?:AM|PM|am|pm)$/;
const TIME_24 = /^\d{1,2}:\d{2}(?::\d{2})?$/;
const YEAR = /^(?:19|20)\d{2}$/;
const FLOAT = /^-?(?:\d+\.\d*|\.\d+)$/;
const INT = /^-?\d+$/;

/**
 * Header labels that mark a column as holding years. Used by the context-aware
 * year resolver (SPEC §5.1.1). Matched case-insensitively as whole words, so
 * "yy-mm" (a month-year date header) deliberately does NOT match.
 */
const YEAR_HEADER = /\b(?:years?|yr|yyyy|fy|fiscal\s*years?)\b/i;

/**
 * Classify a single cell value into a format-aggregation category, by VALUE
 * ALONE. A 4-digit value in 1900–2099 is reported as a `YearData` *candidate*;
 * whether it stays a year or becomes `IntNum` is decided by context in
 * `resolveYear` (SPEC §5.1.1). Returns `null` for the empty string (the only
 * value considered empty per SPEC §3.1).
 */
export function classify(v: string): FormatType | null {
  if (v === "") return null;
  if (BOOLEAN.test(v)) return "Boolean";
  if (EMAIL.test(v)) return "EmailData";
  if (SCIENTIFIC.test(v)) return "ScientificNum";
  if (PERCENT.test(v)) return "PercentageNum";
  if (CURRENCY.test(v)) return "CurrencyData";
  if (DATE_ISO.test(v) || DATE_SLASH.test(v) || DATE_DASH.test(v))
    return "DateData";
  if (TIME_12.test(v) || TIME_24.test(v)) return "TimeData";
  if (YEAR.test(v)) return "YearData";
  if (FLOAT.test(v)) return "FloatNum";
  if (INT.test(v)) return "IntNum";
  return "Text";
}

/**
 * Find the column header governing cell (r, c): the nearest non-empty cell
 * ABOVE it in the same column whose value classifies as Text (a label). Skips
 * blanks and numeric cells so a header above intervening data is still found.
 * Returns null when the column has no text label above the cell.
 */
function nearestHeaderAbove(grid: Grid, r: number, c: number): string | null {
  for (let rr = r - 1; rr >= 0; rr--) {
    const v = grid.rows[rr]?.[c] ?? "";
    if (v === "") continue;
    if (classify(v) === "Text") return v;
  }
  return null;
}

/**
 * Decide whether a year *candidate* at (r, c) is really a `YearData` or an
 * ordinary `IntNum` (SPEC §5.1.1). Priority:
 *   1. Column header is the dominant signal: a year-ish header → YearData; any
 *      other header → IntNum (suppresses a stray in-range integer like a count).
 *   2. No header → column-neighbour signal: stays YearData only if EVERY other
 *      integer-valued cell in the column is also a year (1900–2099) and there is
 *      at least one such neighbour.
 *   3. Isolated in-range integer with no header and no integer neighbours →
 *      IntNum (we don't guess "year" from a lone value).
 */
function resolveYear(grid: Grid, r: number, c: number): FormatType {
  const header = nearestHeaderAbove(grid, r, c);
  if (header !== null) {
    return YEAR_HEADER.test(header) ? "YearData" : "IntNum";
  }

  let intSiblings = 0;
  let yearSiblings = 0;
  const numRows = grid.rows.length;
  for (let rr = 0; rr < numRows; rr++) {
    if (rr === r) continue;
    const t = classify(grid.rows[rr]?.[c] ?? "");
    if (t === "YearData") {
      intSiblings++;
      yearSiblings++;
    } else if (t === "IntNum") {
      intSiblings++;
    }
  }
  if (intSiblings === 0) return "IntNum";
  return yearSiblings === intSiblings ? "YearData" : "IntNum";
}

type Rect = {
  type: FormatType;
  topRow: number;
  leftCol: number;
  bottomRow: number;
  rightCol: number;
};

/**
 * Greedy rectangular aggregation: scan the type map in row-major order; for
 * each unclaimed non-empty cell, extend as far right as the row's same-type
 * run goes, then extend down as long as every row below matches that full
 * width. Mark the rectangle's cells claimed and continue. Empty cells break
 * runs (no aggregation across gaps).
 */
function aggregate(grid: Grid): Rect[] {
  const numRows = grid.rows.length;
  let numCols = 0;
  for (const row of grid.rows) {
    if (row.length > numCols) numCols = row.length;
  }
  if (numRows === 0 || numCols === 0) return [];

  const types: (FormatType | null)[][] = Array.from(
    { length: numRows },
    (_, r) => {
      const row = grid.rows[r] ?? [];
      return Array.from({ length: numCols }, (_, c) => classify(row[c] ?? ""));
    },
  );

  // Context-aware year resolution (SPEC §5.1.1): a value-level YearData stays a
  // year only when its column header / neighbours support it; otherwise IntNum.
  for (let r = 0; r < numRows; r++) {
    const typeRow = types[r]!;
    for (let c = 0; c < numCols; c++) {
      if (typeRow[c] === "YearData") typeRow[c] = resolveYear(grid, r, c);
    }
  }

  const claimed: boolean[][] = Array.from({ length: numRows }, () =>
    new Array<boolean>(numCols).fill(false),
  );

  const rects: Rect[] = [];

  for (let r = 0; r < numRows; r++) {
    const typeRow = types[r]!;
    const claimedRow = claimed[r]!;
    for (let c = 0; c < numCols; c++) {
      if (claimedRow[c]) continue;
      const t = typeRow[c];
      if (t == null) continue;

      // Extend right along row r.
      let w = 1;
      while (c + w < numCols && typeRow[c + w] === t && !claimedRow[c + w]) {
        w++;
      }

      // Extend down: each candidate row must be fully same-type AND unclaimed
      // across the [c, c+w) span.
      let h = 1;
      extendDown: while (r + h < numRows) {
        const nextTypes = types[r + h]!;
        const nextClaimed = claimed[r + h]!;
        for (let cc = c; cc < c + w; cc++) {
          if (nextTypes[cc] !== t || nextClaimed[cc]) break extendDown;
        }
        h++;
      }

      for (let rr = r; rr < r + h; rr++) {
        claimed[rr]!.fill(true, c, c + w);
      }

      rects.push({
        type: t,
        topRow: r,
        leftCol: c,
        bottomRow: r + h - 1,
        rightCol: c + w - 1,
      });
    }
  }

  return rects;
}

function rectToRange(rect: Rect, origin: Origin): string {
  const topLeft = a1(origin.row + rect.topRow, origin.col + rect.leftCol);
  if (rect.topRow === rect.bottomRow && rect.leftCol === rect.rightCol) {
    return topLeft;
  }
  const bottomRight = a1(
    origin.row + rect.bottomRow,
    origin.col + rect.rightCol,
  );
  return `${topLeft}:${bottomRight}`;
}

export function encodeFormatAggregation(
  grid: Grid,
  tokenCounter: TokenCounter,
): Encoding<FormatAggregationJson> {
  const rects = aggregate(grid);

  const byType = new Map<FormatType, string[]>();
  for (const rect of rects) {
    const ranges = byType.get(rect.type) ?? [];
    ranges.push(rectToRange(rect, grid.origin));
    byType.set(rect.type, ranges);
  }

  const groups: FormatAggregationJson["groups"] = [];
  for (const t of TYPE_ORDER) {
    const ranges = byType.get(t);
    if (!ranges || ranges.length === 0) continue;
    groups.push({ type: t, ranges });
  }

  const string = groups
    .map((g) => `${g.type}: ${g.ranges.join(",")}`)
    .join("\n");

  const json: FormatAggregationJson = {
    encoding: "format-aggregation",
    version: 0,
    origin: { row: grid.origin.row, col: grid.origin.col },
    groups,
  };

  return { string, json, tokenEstimate: tokenCounter(string) };
}

File: packages/python/sheet_compressor/encodings/format_aggregation.py
"""Format-aggregation encoding (SPEC §5)."""

import re
from typing import Any, List, Mapping, Optional

from ..address import a1

# SPEC §5.1: canonical emission order.
_TYPE_ORDER = (
    "IntNum",
    "FloatNum",
    "ScientificNum",
    "PercentageNum",
    "CurrencyData",
    "DateData",
    "TimeData",
    "YearData",
    "EmailData",
    "Boolean",
    "Text",
)

# SPEC §5.1 classification patterns, priority order — first match wins.
_BOOLEAN = re.compile(r"^(?:true|false)$", re.IGNORECASE)
_EMAIL = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")
_SCIENTIFIC = re.compile(r"^-?\d+(?:\.\d+)?[eE][+-]?\d+$")
_PERCENT = re.compile(r"^-?\d+(?:\.\d+)?%$")
_CURRENCY = re.compile(r"^-?[$€£¥]\d+(?:\.\d+)?$")
_DATE_ISO = re.compile(r"^\d{4}-\d{1,2}-\d{1,2}$")
_DATE_SLASH = re.compile(r"^\d{1,2}/\d{1,2}/\d{2,4}$")
_DATE_DASH = re.compile(r"^\d{1,2}-\d{1,2}-\d{2,4}$")
_TIME_12 = re.compile(r"^\d{1,2}:\d{2}(?::\d{2})?\s?(?:AM|PM|am|pm)$")
_TIME_24 = re.compile(r"^\d{1,2}:\d{2}(?::\d{2})?$")
_YEAR = re.compile(r"^(?:19|20)\d{2}$")
_FLOAT = re.compile(r"^-?(?:\d+\.\d*|\.\d+)$")
_INT = re.compile(r"^-?\d+$")

# SPEC §5.1.1: column-header labels that mark a column as holding years.
# Matched case-insensitively as whole words.
_YEAR_HEADER = re.compile(r"\b(?:years?|yr|yyyy|fy|fiscal\s*years?)\b", re.IGNORECASE)


def classify(v: str) -> Optional[str]:
    """SPEC §5.1: classify by value alone. Returns None for ""."""
    if v == "":
        return None
    if _BOOLEAN.match(v):
        return "Boolean"
    if _EMAIL.match(v):
        return "EmailData"
    if _SCIENTIFIC.match(v):
        return "ScientificNum"
    if _PERCENT.match(v):
        return "PercentageNum"
    if _CURRENCY.match(v):
        return "CurrencyData"
    if _DATE_ISO.match(v) or _DATE_SLASH.match(v) or _DATE_DASH.match(v):
        return "DateData"
    if _TIME_12.match(v) or _TIME_24.match(v):
        return "TimeData"
    if _YEAR.match(v):
        return "YearData"
    if _FLOAT.match(v):
        return "FloatNum"
    if _INT.match(v):
        return "IntNum"
    return "Text"


def _cell(rows, r: int, c: int) -> str:
    if r < 0 or r >= len(rows):
        return ""
    row = rows[r]
    if c < 0 or c >= len(row):
        return ""
    return row[c]


def _nearest_header_above(rows, r: int, c: int) -> Optional[str]:
    """SPEC §5.1.1: nearest Text-classified cell above (r, c) in the same column."""
    rr = r - 1
    while rr >= 0:
        v = _cell(rows, rr, c)
        if v == "":
            rr -= 1
            continue
        if classify(v) == "Text":
            return v
        rr -= 1
    return None


def _resolve_year(rows, r: int, c: int) -> str:
    """SPEC §5.1.1: a year *candidate* (already matched _YEAR) is YearData iff
    the column context supports it; otherwise IntNum.
    """
    header = _nearest_header_above(rows, r, c)
    if header is not None:
        return "YearData" if _YEAR_HEADER.search(header) else "IntNum"

    int_siblings = 0
    year_siblings = 0
    for rr in range(len(rows)):
        if rr == r:
            continue
        t = classify(_cell(rows, rr, c))
        if t == "YearData":
            int_siblings += 1
            year_siblings += 1
        elif t == "IntNum":
            int_siblings += 1
    if int_siblings == 0:
        return "IntNum"
    return "YearData" if year_siblings == int_siblings else "IntNum"


def _aggregate(grid: Mapping[str, Any]):
    rows = grid.get("rows", [])
    num_rows = len(rows)
    num_cols = 0
    for row in rows:
        if len(row) > num_cols:
            num_cols = len(row)
    if num_rows == 0 or num_cols == 0:
        return []

    types: List[List[Optional[str]]] = []
    for r in range(num_rows):
        row = rows[r]
        type_row: List[Optional[str]] = []
        for c in range(num_cols):
            value = row[c] if c < len(row) else ""
            type_row.append(classify(value))
        types.append(type_row)

    # SPEC §5.1.1: resolve YearData candidates by column context.
    for r in range(num_rows):
        for c in range(num_cols):
            if types[r][c] == "YearData":
                types[r][c] = _resolve_year(rows, r, c)

    claimed = [[False] * num_cols for _ in range(num_rows)]

    rects = []
    for r in range(num_rows):
        for c in range(num_cols):
            if claimed[r][c]:
                continue
            t = types[r][c]
            if t is None:
                continue

            # Extend right along row r.
            w = 1
            while (
                c + w < num_cols
                and types[r][c + w] == t
                and not claimed[r][c + w]
            ):
                w += 1

            # Extend down: each candidate row must be fully same-type AND
            # unclaimed across [c, c+w).
            h = 1
            while r + h < num_rows:
                ok = True
                for cc in range(c, c + w):
                    if types[r + h][cc] != t or claimed[r + h][cc]:
                        ok = False
                        break
                if not ok:
                    break
                h += 1

            for rr in range(r, r + h):
                for cc in range(c, c + w):
                    claimed[rr][cc] = True

            rects.append((t, r, c, r + h - 1, c + w - 1))

    return rects


def _rect_to_range(rect, origin) -> str:
    t, top, left, bottom, right = rect
    top_left = a1(origin["row"] + top, origin["col"] + left)
    if top == bottom and left == right:
        return top_left
    bottom_right = a1(origin["row"] + bottom, origin["col"] + right)
    return f"{top_left}:{bottom_right}"


def encode_format_aggregation(grid: Mapping[str, Any], token_counter) -> dict:
    rects = _aggregate(grid)
    origin = grid["origin"]

    by_type: "dict[str, list[str]]" = {}
    for rect in rects:
        ranges = by_type.setdefault(rect[0], [])
        ranges.append(_rect_to_range(rect, origin))

    groups = []
    for t in _TYPE_ORDER:
        ranges = by_type.get(t)
        if not ranges:
            continue
        groups.append({"type": t, "ranges": ranges})

    string = "\n".join(f"{g['type']}: {','.join(g['ranges'])}" for g in groups)

    json_obj = {
        "encoding": "format-aggregation",
        "version": 0,
        "origin": {"row": origin["row"], "col": origin["col"]},
        "groups": groups,
    }
    return {"string": string, "json": json_obj, "tokenEstimate": token_counter(string)}
